### ISSUE TITLE
Use findOne (or find) does not return expected fields

### ISSUE THREAD

**Issue by EmmanuelVella** [alcaeus/mongo-php-adapter#144]
Hi !

Doing `MongoGridFS::findOne('58666125d77f1.png')` returns all fields as expected. For example :

```
MongoGridFSFile {#2723 ▼
  +file: array:8 [▼
    "_id" => MongoId {#2549 ▶}
    "date" => MongoDate {#2721 ▶}
    "contentType" => "image/png"
    "filename" => "58666125d77f1.png"
    "length" => 401381
    "md5" => "c1be06a1dee81eb405e4df4fb096ba28"
    "uploadDate" => MongoDate {#2724 ▶}
    "chunkSize" => 261120
  ]
  #gridfs: MongoGridFS {#1348 ▶}
}
```

However, doing `MongoGridFS::findOne('58666125d77f1.png', array('date'))` returns only the `_id` field. The expected `date` field is not returned :

```
GridFS.php on line 150:
MongoGridFSFile {#2725 ▼
  +file: array:1 [▼
    "_id" => MongoId {#2549 ▶}
  ]
  #gridfs: MongoGridFS {#1348 ▶}
}
```

I'm using version 1.0.7. 

Thank you !

**Comment by alcaeus** [comment]
Good catch. I tried reproducing it and was about to comment that it isn't a bug (since projection syntax is `findOne('58666125d77f1.png', array('date' => true))`. I then realized that this was never documented but indeed done in the legacy driver: https://github.com/mongodb/mongo-php-driver-legacy/blob/e6f687cd12dea23da23ca8a9a9a98eda1e0e0ecc/cursor.c#L107..L112

This will be fixed in 1.0.8, thanks for the report!

**Comment by EmmanuelVella** [comment]
Thank you very much ! 👍

**Comment by palmic** [comment]
You broke previous functionality by this fix.
This worked before:
`$collection->find($query, ['field' => 1])`

Now it returns empty document.
The code has to be updated to this:
`$collection->find($query, ['_id' => true, 'field' => 1])`

You should not add such updates into minor version, now i am forced to write integration tests just to make sure your module work still the same after minor update.

**Comment by alcaeus** [comment]
@palmic if you can provide a failing test case, I'll be glad to take a look. FWIW, the following test case (introduced with #156) provides the expected result:
```php
    public function testFindWithProjectionWithInt()
    {
        $document = ['foo' => 'foo', 'bar' => 'bar'];
        $this->getCollection()->insert($document);
        unset($document['_id']);
        $this->getCollection()->insert($document);

        $cursor = $this->getCollection()->find(['foo' => 'foo'], ['bar' => 1]);
        foreach ($cursor as $document) {
            $this->assertCount(2, $document);
            $this->assertArrayHasKey('_id', $document);
            $this->assertArraySubset(['bar' => 'bar'], $document);
        }
    }
```

Looking at it from a logical standpoint, it doesn't make any sense either. The `_id` field has to be explicitly excluded by passing `['_id' => false]` as projection. Just omitting the field will not skip it in any case.

**Comment by palmic** [comment]
Sorry, it was not so simple to have any fields in doc.
[Check this](https://github.com/palmic/mongodb-php-test) and thanks for so quick response!

**Comment by alcaeus** [comment]
Thank you for the test case. The issue is that you are passing integers as keys for the projection array, which wasn't the case in the sample code you provided before. This is an edge case that only partially works in the legacy driver. I've created three tests in a branch: https://github.com/alcaeus/mongo-php-adapter/compare/1.0.x...projection-with-numeric-keys
Check the build result for the control build (tests are run against ext-mongo): https://travis-ci.org/alcaeus/mongo-php-adapter/jobs/195873254#L1059

Field names in MongoDB are strings, not integers, so in your projection you should pass them as strings. When handling query projections, the legacy driver checks for the `fields` parameter being a numeric array, (sequential indexes starting at 0). If so, it switches to the legacy projection mode also presented in this issue. That projection mode also requires strings for field names, which the other checks don't.

For the record, you might run into a bunch of other issues by using numeric keys in documents, so I'd suggest casting them to string. This is due to PHP not differentiating between hashes and arrays, requiring libraries to try and detect arrays by checking for numeric keys, with better or worse luck.

**Comment by palmic** [comment]
Hm Understand now, thanks you for great explanation and your time.
Michal.

**Comment by alcaeus** [comment]
You're welcome. I'll see if I can easily fix the regression introduced in #145.

**Comment by palmic** [comment]
Ok, just please take care to change API behaviour in minor version ;-)
As i understand to your update and unfortunate occasions about this issue, this really broke our app because our invalid usage of this.

Thank you for your work!

**Comment by alcaeus** [comment]
Since the original change was considered a bug and I also consider this regression a bug, it would be fixed in 1.0.9, as is perfectly acceptable in SemVer.

**Comment by palmic** [comment]
Thank you!
